server/src/main/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticator.java

```diff
@@ -148,8 +149,20 @@ private void validateClaims(final Map<String, Object> claims) {
         }
 
         // require audience claim
-        if (!StringUtils.hasText((String) claims.get(ClaimConstants.AUD))) {
-            throw new InvalidTokenException("aud claim is required in jwt-bearer assertion.");
+        Object audience = claims.get(ClaimConstants.AUD);
+        if (audience instanceof ArrayList) {
+            ArrayList<?> audienceList = (ArrayList<?>) audience;
+            if (audienceList.isEmpty()) {
+                throw new InvalidTokenException("aud claim is required in jwt-bearer assertion.");
+            }
+        }
+        else if(audience instanceof String) {
+            if (!StringUtils.hasText((String) claims.get(ClaimConstants.AUD))) {
+                throw new InvalidTokenException("aud claim is required in jwt-bearer assertion.");
+            }
+        }
+        else {
+            throw new InvalidTokenException("aud claim is in the wrong format.");
```

Comment:
In light of expense related to Reflection(instanceof) and Exceptions, is there a real value in doing this twice?
Probably the validateClaims should just be checking **the existence(not-null) of **audience** claim**, and leave the **detailed validation** to the assertAudience

Reply:
I see your point, but validateClaims function is checking that a value exists for audience. If it is an array, just checking that the object exists does not rule out an empty array. The assertAudience function is checking that the issuerUrl matches what is in the audience claim. So these two functions have two different jobs based on how we previously wrote the code.

Reply:
We could possibly improve the performance by substituting the value in the Claims map for the key:` ClaimConstants.AUD)`
with an array, so that in the `assertAudience`  only the `audienceList.contains(issuerURL)` check is required.
Just a suggestion

Reply:
The only reason I don't think it should be pushed to claimsMap function (as this is the function that builds the map in the first place) is that all those checks and exceptions would be pushed there. If we modify the map after we build it then why don't we build the map correctly the first time. Then the claimsMap function logically wouldn't follow its naming. Not only would it be doing mapping, but it would also be doing the checking for audience only.  I don't think there will be a noticeable performance impact on this check.

Reply:
Casting and instanceOf TWO additional times is not going to be a noticeable impact.

On a casual note the claimsMap function is a wrapper on JsonUtils.readValue, which is being done several times.
More importantly both validateClaims() and assertAudience() are being called on the same local claims variable in the method----authenticateJwtAssertionToken().

Reply:
Looks good overall